File: frontend/src/types/prediction-market.ts
export type PredictionMarketMode = "paper";
export type PredictionMarketFreshnessStatus =
  | "fresh"
  | "delayed"
  | "stale"
  | "unavailable";

export interface PredictionMarketOutcome {
  outcome: string;
  token_id: string;
  price?: string;
}

export interface PredictionMarketSummary {
  market_id: string;
  slug: string;
  question: string;
  active: boolean;
  closed: boolean;
  outcomes: PredictionMarketOutcome[];
}

export interface PredictionMarketCatalog {
  source: "polymarket-public";
  mode: PredictionMarketMode;
  source_timestamp_ms: number;
  observed_at_ms: number;
  freshness_age_ms: number;
  freshness_status: PredictionMarketFreshnessStatus;
  markets: PredictionMarketSummary[];
}

export interface PredictionMarketBookLevel {
  price: string;
  size: string;
}

export interface PredictionMarketBookHealth {
  status?: string;
  reason?: string;
  crossed?: boolean;
  one_sided?: boolean;
  bid_levels?: number;
  ask_levels?: number;
}

export interface PredictionMarketOrderBook {
  bids: PredictionMarketBookLevel[];
  asks: PredictionMarketBookLevel[];
  best_bid?: string;
  best_ask?: string;
  midpoint?: string;
  microprice?: string;
  health?: PredictionMarketBookHealth;
}

export interface PredictionMarketSignal {
  reference_price?: string;
  reference_method?: string;
  volatility?: string;
  observation_count?: number;
  volatility_status?: string;
}

export interface PredictionMarketSnapshot {
  source: "polymarket-public";
  mode: PredictionMarketMode;
  source_timestamp_ms: number;
  observed_at_ms: number;
  freshness_age_ms: number;
  freshness_status: PredictionMarketFreshnessStatus;
  market_id: string;
  question: string;
  outcome: string;
  token_id: string;
  book: PredictionMarketOrderBook;
  signal?: PredictionMarketSignal;
}

export interface PredictionReplayRequest {
  decision_time_ms: number;
  latency_ms: number;
  order: {
    side: "buy" | "sell";
    size: number;
    max_levels: number;
    extra_slippage_bps: number;
  };
  snapshots: Array<{
    source_timestamp_ms: number;
    observed_at_ms: number;
    bids: PredictionMarketBookLevel[];
    asks: PredictionMarketBookLevel[];
  }>;
}

export interface PredictionReplayResult {
  source: "polymarket-public";
  mode: "paper";
  simulation_mode: "simulated";
  source_timestamp_ms: number | null;
  observed_at_ms: number | null;
  freshness_age_ms: number | null;
  freshness_status: PredictionMarketFreshnessStatus;
  fingerprint: string;
  assumptions: {
    eligible_time_ms: number;
    execution_snapshot_timestamp_ms: number | null;
    max_levels: number;
    extra_slippage_bps: number;
    remainder_policy: "cancel";
    liquidity_scope: "visible_frozen_levels";
  };
  fill: {
    requested_size: number;
    filled_size: number;
    unfilled_size: number;
    vwap: number | null;
    levels_consumed: number;
  };
  mark_to_book: {
    mark_price: number | null;
    pnl: number;
    currency: "quote";
  };
}

File: python/valuecell/server/api/schemas/prediction_market.py
"""Public Polymarket research API contracts.

These contracts deliberately expose public market observations and paper-only
research data. They contain no credential, wallet, or order-execution field.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from .base import SuccessResponse


class PredictionMarketOutcomeData(BaseModel):
    outcome: str
    token_id: str
    price: str | None = None


class PredictionMarketSummaryData(BaseModel):
    market_id: str
    slug: str = ""
    question: str
    active: bool
    closed: bool
    outcomes: list[PredictionMarketOutcomeData]


class PredictionMarketCatalogData(BaseModel):
    source: Literal["polymarket-public"] = "polymarket-public"
    mode: Literal["paper"] = "paper"
    source_timestamp_ms: int
    observed_at_ms: int
    freshness_age_ms: int
    freshness_status: Literal["fresh", "delayed", "stale", "unavailable"]
    markets: list[PredictionMarketSummaryData]
    next_cursor: str | None = None
    warnings: list[str] = Field(default_factory=list)


class PredictionMarketBookLevelData(BaseModel):
    price: str
    size: str


class PredictionMarketBookHealthData(BaseModel):
    status: Literal["valid", "crossed", "one_sided", "empty", "invalid", "stale"]
    reason: str | None = None
    crossed: bool = False
    one_sided: bool = False
    bid_levels: int = 0
    ask_levels: int = 0


class PredictionMarketOrderBookData(BaseModel):
    bids: list[PredictionMarketBookLevelData]
    asks: list[PredictionMarketBookLevelData]
    best_bid: str | None = None
    best_ask: str | None = None
    midpoint: str | None = None
    microprice: str | None = None
    health: PredictionMarketBookHealthData


class PredictionMarketSignalData(BaseModel):
    reference_price: str | None = None
    reference_method: Literal["microprice", "midpoint", "unavailable"] = "unavailable"
    volatility: str | None = None
    observation_count: int = 0
    volatility_status: Literal["available", "insufficient_history", "invalid_history"]


class PredictionMarketSnapshotData(BaseModel):
    source: Literal["polymarket-public"] = "polymarket-public"
    mode: Literal["paper"] = "paper"
    source_timestamp_ms: int
    observed_at_ms: int
    freshness_age_ms: int
    freshness_status: Literal["fresh", "delayed", "stale", "unavailable"]
    market_id: str
    question: str
    outcome: str
    token_id: str
    book: PredictionMarketOrderBookData
    signal: PredictionMarketSignalData | None = None
    warnings: list[str] = Field(default_factory=list)


PredictionMarketCatalogResponse = SuccessResponse[PredictionMarketCatalogData]
PredictionMarketSnapshotResponse = SuccessResponse[PredictionMarketSnapshotData]
